### ISSUE TITLE
Bug: Distinct() does not work for complex types 

### ISSUE THREAD

**Issue by kutyel** [kutyel/linq.ts#52]
The same test case of `DistinctBy` should pass without providing any key selector, I wanted to refactor the **Distinct** function some time ago using this trick:

```js
return new List([...new Set(this._elements)])
```

Nevertheless, this does not work either, help would be much appreciated.

cc: @abbasmhd @IngoBleile

**Comment by fredrik-lundin** [comment]
I guess this doesn't work since `Set` is using `===` internally, and `{ a: "a", b: "b" } === { a: "a", b: "b" }` is false.

To just make it work, this code should do it (haven't done a lot of testing though). But order of the properties are important using this strategy. Methods on the objects are also not allowed for this to work.

```javascript
return new List( [...new Set( arr.map( obj => JSON.stringify( obj ) ) )].map( str => JSON.parse( str ) ) );
```

**Comment by vendethiel** [comment]
Comparing JSON encodings is **really** slow, and will not show all properties.

**Comment by fredrik-lundin** [comment]
@vendethiel, i tried it with an array of 1 000 000 objects, and that took less than 2 seconds in my browser. So yea, not blazing fast, but I don't know about *really* slow.

**Comment by vendethiel** [comment]
That's insanely slow. Dealing with a million of objects isn't "out of this world", I've had to work on applications that did rendering with more (no, no way to reduce the number the number of things...), and this was the major bottleneck when doing update diffing. Using a recursive comparison strategy (like lodash / underscore have) was several orders of magnitude fasters.

**Comment by fredrik-lundin** [comment]
@vendethiel, you are absolutely right!

**Comment by kutyel** [comment]
What about this comparison, would it be more performant?

```typescript
    /**
     * Determine if two objects are equal
     */
    private _equals<T, U>(x: T, y: U): boolean {
        const keys = Object.keys(x);
        return keys.length === Object.keys(y).length && keys.every(k => this[k] === y[k]);
    }
```

cc: @fredrik-lundin @vendethiel

**Comment by rhyek** [comment]
What about using [deep-equal](https://www.npmjs.com/package/deep-equal)?

**Comment by vendethiel** [comment]
@kutyel that wouldn't be deep

**Comment by kutyel** [comment]
You are all welcome to provide a PR if you feel like contributing to the
project 😉

**Comment by rhyek** [comment]
@kutyel: I could provide a PR, but before I work on it I'd like to know your thoughts. IMO, `Distinct` and `DistinctBy` should use `GroupBy` internally and just return the result's keys. `GroupBy` should return a `Map` so keys can be complex objects. The return type for `GroupBy` would need to change and that would probably mean fixing #56 as well.

If that sounds OK to you I'll go ahead and work on it.

**Comment by kutyel** [comment]
Sounds great! Thanks for your hard work! 👍🏼

**Comment by isc30** [comment]
Hi all, it's **9 months** since the bug was opened and we still don't have a fix.
Time to switch to a **better lib**? Check ES5 compatible https://github.com/isc30/linq-collections

**Comment by vendethiel** [comment]
You're an ass.

**Comment by isc30** [comment]
Check this, its **exactly** what he did some hours ago
https://github.com/isc30/linq-collections/issues/11

**Comment by vendethiel** [comment]
Awful :|.

**Comment by isc30** [comment]
i know

**Comment by kutyel** [comment]
Hi @rhyek! Any advance on this? I was going to fixit myself now but would not want to override your work if you have any progress on this 😉

**Comment by rhyek** [comment]
Hey, @kutyel. I actually did do the fix for this, but it was during a very busy time for me and was having issues with the repo's precommit scripts (not 100% sure).

Some time later I built this https://github.com/rhyek/group-by-object. You mind find it or its implementation details useful.

Sorry!

**Comment by kutyel** [comment]
This should do the trick https://github.com/kutyel/linq.ts/commit/d9346e61df0714cc2e64229bc68167d8c2db5859 but it's not working for some reason, any advice is welcome people! 😉  cc: @rhyek @vendethiel @fredrik-lundin

**Comment by vendethiel** [comment]
The comments on `_equals` not being recursive stands in that commit

**Comment by kutyel** [comment]
True, will work in improving that function and make it deep, although a PR would be also greatly welcome 😉

**Comment by kutyel** [comment]
I am super glad to announce this will be fixed in the next release! 🎉 🎉 🎉 See: 153a37d792c28535826cd1b0395a36a10d493a83
